h5py/_hl/dataset.py

```diff
@@ -173,6 +175,9 @@ def make_new_dset(parent, shape=None, dtype=None, data=None, name=None,
         sid = h5s.create_simple(shape, maxshape)
 
     dset_id = h5d.create(parent.id, name, tid, sid, dcpl=dcpl, dapl=dapl)
+    # Multiple NumPy dtypes can map to the same hdf5 type; e.g.
+    # object strings or NpyStrings map to vlen strings.
+    dset_id._dtype = dtype
```

Comment:
Let's avoid this - a newly created dataset should not behave any differently from one retrieved from an existing file. I.e. it should not know that it was created with the new string dtype.

Reply:
This is what makes it possible to write
```python
f.create_dataset("x", ["foo"], dtype="T")
```
or
```python
data = np.asarray(["foo"], dtype="T")
f.create_dataset("x", data)
```
otherwise, the dtype information would be immediately forgotten and I think you'd agree it would be mightily confusing for end users to have to call `.astype("T")` immediately afterwards.